File: DSA/DivyanshKumarSingh_590012175/Day_12_Question_3.py
def compress(s):
    if not s:
        return ""
    
    compressed = []
    c = 1

    for i in range(1, len(s)):
        if s[i] == s[i - 1]:
            c += 1
        else:
            compressed.append(s[i - 1] + str(c))
            c = 1

    compressed.append(s[-1] + str(c))

    cs = ''.join(compressed)

    
    return cs if len(cs) <= len(s) else s

inp = input("Enter a string: ").strip()
print("Compressed String:", compress(inp))

File: DSA/PriyankaSingh_590016827/day12question3.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

char* compressString(const char* str) 
{
    int len = strlen(str);
    if (len == 0) return strdup("");
    
    // Allocating maximum possible compressed string length (2x original)
    char* compressed = malloc(2 * len + 1);
    if (!compressed) return NULL;
    
    int count = 1;
    int index = 0;
    
    for (int i = 1; i <= len; i++) 
    {
        if (i < len && str[i] == str[i-1]) 
        {
            count++;
        } 
        else 
        {
            // Append character and count
            compressed[index++] = str[i-1];
            int num_chars = snprintf(compressed + index, 10, "%d", count);
            index += num_chars;
            count = 1;
        }
    }
    
    compressed[index] = '\0';
    
    // Returning original if compression didn't help
    if (strlen(compressed) >= len) 
    {
        free(compressed);
        return strdup(str);
    }
    
    return compressed;
}

int main() 
{
    char input[100];
    
    printf("Enter a string to compress: ");
    fgets(input, sizeof(input), stdin);
    input[strcspn(input, "\n")] = '\0'; // Removing newline
    
    char* compressed = compressString(input);
    
    printf("Compressed string: %s\n", compressed);
    
    free(compressed);
    return 0;
}